### ISSUE TITLE
Inconsistent result in calculations with Dynamic field construction and destruction

### ISSUE THREAD

**Issue by ifsheldon** [taichi-dev/taichi#4334]
**Describe the bug**
Running code multiple times, in which fields are dynamically constructed and destructed, will have different output, which may be caused by memory problems. The inconsistency results in valid numbers and sometimes Nan values.

**To Reproduce**
Try this test case
```python
def test_vvf_ti():
    ti.init(ti.cpu)
    for i in range(10):
        print(i)
        input_arguments_list = torch.load("input_argument_list.pt")
        result = vvf_interpolate(*input_arguments_list)
        assert not torch.any(torch.isnan(result)) # first multiple iterations are fine
```
And the `vvd_interpolate` is defined as 
```python
import taichi as ti
import torch

from taichi.lang.util import to_taichi_type

@ti.func
def cast_int(x):
    return ti.cast(x, ti.i32)


@ti.kernel
def interpolate_along_y_kernel(grid_point_values: ti.template(),
                               grid_coord_along_axis: ti.template(),
                               point_pos: ti.template(),
                               order_along_y: ti.template(),
                               dy: ti.template(),
                               predecessor_pos: ti.template(),
                               intervals: ti.template(),
                               output_values: ti.template()):
    for point in range(point_pos.shape[0]):
        for ax in range(grid_coord_along_axis.shape[0] - 1):
            interval_start = intervals[point, 0]  # interval_start is floor(predecessor_location)
            interval_end = intervals[point, 1]  # interval_end is floor(point_location)
            is_first = order_along_y[point] == 0
            if not is_first:
                if interval_start <= interval_end:
                    if interval_start <= ax <= interval_end:
                        x = point_pos[point, 0]
                        y = point_pos[point, 1]
                        belong_x_ax = cast_int(x)
                        p_y = predecessor_pos[point, 1]
                        if interval_start == interval_end:
                            # now interval_start == interval_end == ax == floor(x)
                            value = (y - p_y) / dy[None] * grid_point_values[belong_x_ax, ax]
                            output_values[point] += value
                        else:
                            if ax == interval_start:
                                stride = grid_coord_along_axis[ax + 1] - p_y
                                value = stride / dy[None] * grid_point_values[belong_x_ax, ax]
                                output_values[point] += value
                            elif ax == interval_end:
                                stride = y - grid_coord_along_axis[ax]
                                value = stride / dy[None] * grid_point_values[belong_x_ax, ax]
                                output_values[point] += value
                            else:
                                value = (grid_coord_along_axis[ax + 1] - grid_coord_along_axis[ax]) / dy[None] \
                                        * grid_point_values[belong_x_ax, ax]
                                output_values[point] += value
                else:
                    if interval_start <= ax:
                        x = point_pos[point, 0]
                        belong_x_ax = cast_int(x)
                        p_y = predecessor_pos[point, 1]
                        if ax == interval_start:
                            stride = grid_coord_along_axis[ax + 1] - p_y
                            value = stride / dy[None] * grid_point_values[belong_x_ax, ax]
                            output_values[point] += value
                        elif ax < grid_coord_along_axis.shape[0] - 1:
                            value = (grid_coord_along_axis[ax + 1] - grid_coord_along_axis[ax]) / dy[None] \
                                    * grid_point_values[belong_x_ax, ax]
                            output_values[point] += value
                        else:
                            pass


def create_field_from_tensor(tensor):
    dtype = tensor.dtype
    if dtype == torch.cdouble:
        dtype = torch.double
        x = ti.Vector.field(2, to_taichi_type(dtype))
    elif dtype == torch.cfloat:
        dtype = torch.float
        x = ti.Vector.field(2, to_taichi_type(dtype))
    else:
        x = ti.field(to_taichi_type(tensor.dtype))

    fb = ti.FieldsBuilder()
    fb.dense(ti.axes(*range(len(tensor.shape))), tensor.shape).place(x)
    snode_tree = fb.finalize()
    return x, snode_tree


def vvf_interpolate(grid_point_values: torch.Tensor,
                    grid_coord_along_axis: torch.Tensor,
                    point_pos: torch.Tensor,
                    order_along_axis: torch.Tensor,
                    dy: torch.Tensor,
                    predecessor_pos: torch.Tensor,
                    intervals: torch.Tensor,
                    axis: str):
    kernel = interpolate_along_y_kernel

    grid_point_values_field, s1 = create_field_from_tensor(grid_point_values)
    grid_coord_along_axis_field, s2 = create_field_from_tensor(grid_coord_along_axis)
    point_pos_field, s3 = create_field_from_tensor(point_pos)
    order_along_axis_field, s4 = create_field_from_tensor(order_along_axis)
    dy_field, s5 = create_field_from_tensor(dy)
    predecessor_pos_field, s6 = create_field_from_tensor(predecessor_pos)
    intervals_field, s7 = create_field_from_tensor(intervals)
    output_tensor = torch.zeros(point_pos.shape[0], dtype=grid_point_values.dtype)
    output_field, s8 = create_field_from_tensor(output_tensor)

    grid_point_values_field.from_torch(torch.view_as_real(grid_point_values))
    grid_coord_along_axis_field.from_torch(grid_coord_along_axis)
    point_pos_field.from_torch(point_pos)
    order_along_axis_field.from_torch(order_along_axis)
    dy_field.from_torch(dy)
    predecessor_pos_field.from_torch(predecessor_pos)
    intervals_field.from_torch(intervals)

    kernel(grid_point_values_field, grid_coord_along_axis_field, point_pos_field, order_along_axis_field, dy_field,
           predecessor_pos_field, intervals_field, output_field)
    output = output_field.to_torch()
    for s in [s1, s2, s3, s4, s5, s6, s7, s8]:
        s.destroy()

    return output

```

**Log/Screenshots**
Please post the **full log** of the program (instead of just a few lines around the error message, unless the log is > 1000 lines). This will help us diagnose what's happening. For example:

```
$ python my_sample_code.py
test_vector_valued_function.py::test_vvf_ti FAILED                       [100%][Taichi] Starting on arch=arm64
0
1
2
3
4

test_vector_valued_function.py:188 (test_vvf_ti)
def test_vvf_ti():
        ti.init(ti.cpu)
        for i in range(10):
            print(i)
            input_arguments_list = torch.load("input_argument_list.pt")
            result = vvf_interpolate(*input_arguments_list)
>           assert not torch.any(torch.isnan(result))
E           assert not tensor(True)
E            +  where tensor(True) = <built-in method any of type object at 0x1376a80a8>(tensor([[False, False],\n        [False, False],\n        [False, False],\n        ...,\n        [False, False],\n        [False, False],\n        [False, False]]))
E            +    where <built-in method any of type object at 0x1376a80a8> = torch.any
E            +    and   tensor([[False, False],\n        [False, False],\n        [False, False],\n        ...,\n        [False, False],\n        [False, False],\n        [False, False]]) = <built-in method isnan of type object at 0x1376a80a8>(tensor([[0.0000, 0.0000],\n        [0.8674, 0.0000],\n        [0.8675, 0.0000],\n        ...,\n        [2.6174, 0.0000],\n        [2.6174, 0.0000],\n        [2.6174, 0.0000]]))
E            +      where <built-in method isnan of type object at 0x1376a80a8> = torch.isnan

test_vector_valued_function.py:195: AssertionError
```

**Additional comments**
`ti diagnose` results
```
[Taichi] version 0.8.11, llvm 10.0.0, commit 9d0973e5, osx, python 3.8.12

*******************************************
**      Taichi Programming Language      **
*******************************************

Docs:   https://docs.taichi.graphics/
GitHub: https://github.com/taichi-dev/taichi/
Forum:  https://forum.taichi.graphics/

Taichi system diagnose:

python: 3.8.12 | packaged by conda-forge | (default, Jan 30 2022, 23:13:24) 
[Clang 11.1.0 ]
system: darwin
executable: /Users/ifsheldon/miniforge3/bin/python
platform: macOS-12.3-arm64-arm-64bit
architecture: 64bit 
uname: uname_result(system='Darwin', node='maplemac.kaust.edu.sa', release='21.4.0', version='Darwin Kernel Version 21.4.0: Sat Jan 29 03:25:23 PST 2022; root:xnu-8020.100.417.0.4~31/RELEASE_ARM64_T8101', machine='arm64', processor='arm')
locale: zh_CN.UTF-8
PATH: /Users/ifsheldon/miniforge3/bin:/Users/ifsheldon/miniforge3/condabin:/Users/ifsheldon/flutter/bin:/opt/homebrew/bin:/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin:/usr/local/share/dotnet:~/.dotnet/tools:/Library/Apple/usr/bin:/Library/Frameworks/Mono.framework/Versions/Current/Commands:/Users/ifsheldon/.cargo/bin
PYTHONPATH: ['/Users/ifsheldon/miniforge3/bin', '/Users/ifsheldon/miniforge3/lib/python38.zip', '/Users/ifsheldon/miniforge3/lib/python3.8', '/Users/ifsheldon/miniforge3/lib/python3.8/lib-dynload', '/Users/ifsheldon/miniforge3/lib/python3.8/site-packages']

`lsb_release` not available: Command '['lsb_release', '-a']' died with <Signals.SIGTRAP: 5>.


import: ERROR Command '['/Users/ifsheldon/miniforge3/bin/python', '-c', 'import taichi as ti; print("===="); print(ti, end="")']' died with <Signals.SIGTRAP: 5>.

cc: ERROR Command '['/Users/ifsheldon/miniforge3/bin/python', '-c', 'import taichi as ti; print("===="); print(ti.is_arch_supported(ti.cc), end="")']' died with <Signals.SIGTRAP: 5>.
cpu: ERROR Command '['/Users/ifsheldon/miniforge3/bin/python', '-c', 'import taichi as ti; print("===="); print(ti.is_arch_supported(ti.cpu), end="")']' died with <Signals.SIGTRAP: 5>.
metal: ERROR Command '['/Users/ifsheldon/miniforge3/bin/python', '-c', 'import taichi as ti; print("===="); print(ti.is_arch_supported(ti.metal), end="")']' died with <Signals.SIGTRAP: 5>.
opengl: ERROR Command '['/Users/ifsheldon/miniforge3/bin/python', '-c', 'import taichi as ti; print("===="); print(ti.is_arch_supported(ti.opengl), end="")']' died with <Signals.SIGTRAP: 5>.
cuda: ERROR Command '['/Users/ifsheldon/miniforge3/bin/python', '-c', 'import taichi as ti; print("===="); print(ti.is_arch_supported(ti.cuda), end="")']' died with <Signals.SIGTRAP: 5>.
vulkan: ERROR Command '['/Users/ifsheldon/miniforge3/bin/python', '-c', 'import taichi as ti; print("===="); print(ti.is_arch_supported(ti.vulkan), end="")']' died with <Signals.SIGTRAP: 5>.

`glewinfo` not available: Command '['glewinfo']' died with <Signals.SIGTRAP: 5>.

`nvidia-smi` not available: Command '['nvidia-smi']' died with <Signals.SIGTRAP: 5>.
`import taichi` failed: Command '['/Users/ifsheldon/miniforge3/bin/python', '-c', 'import taichi']' died with <Signals.SIGTRAP: 5>.
`ti.init()` failed: Command '['/Users/ifsheldon/miniforge3/bin/python', '-c', 'import taichi as ti; ti.init()']' died with <Signals.SIGTRAP: 5>.
Taichi OpenGL test failed: Command '['/Users/ifsheldon/miniforge3/bin/python', '-c', 'import taichi as ti; ti.init(arch=ti.opengl)']' died with <Signals.SIGTRAP: 5>.
Taichi CUDA test failed: Command '['/Users/ifsheldon/miniforge3/bin/python', '-c', 'import taichi as ti; ti.init(arch=ti.cuda)']' died with <Signals.SIGTRAP: 5>.
`python/taichi/examples/algorithm/laplace.py` failed: Command '['/Users/ifsheldon/miniforge3/bin/python', '-m', 'taichi', 'example', 'minimal']' died with <Signals.SIGTRAP: 5>.
Consider attaching this log when maintainers ask about system information.
>>> Running time: 0.02s

```

My manually compiled version at a7f09f27 also has this problem.

**Comment by ifsheldon** [comment]
And the needed `input_argument_list.pt` is here
[input_argument_list.pt.zip](https://github.com/taichi-dev/taichi/files/8102252/input_argument_list.pt.zip)
.

**Comment by ifsheldon** [comment]
If you comment out 
```python
for s in [s1, s2, s3, s4, s5, s6, s7, s8]:
        s.destroy()
```
The problem is gone, but it's not how problems get solved.

**Comment by k-ye** [comment]
Did we forget to reset memory to zero after the SNodeTree is destroyed (or a new SNodeTree is created)?

**Comment by lin-hitonami** [comment]
> And the needed `input_argument_list.pt` is here [input_argument_list.pt.zip](https://github.com/taichi-dev/taichi/files/8102252/input_argument_list.pt.zip) .

This file seems to be encrypted... I need the password to open it.

**Comment by ifsheldon** [comment]
> This file seems to be encrypted... I need the password to open it.

Sorry, I forgot that my default setting of 7 zip is encrypting with password. Here is the data.
[input_argument_list.pt.zip](https://github.com/taichi-dev/taichi/files/8115679/input_argument_list.pt.zip)

**Comment by strongoier** [comment]
Is this problem solved? @lin-hitonami

**Comment by lin-hitonami** [comment]
Yes, it is solved by #4463.
